### PR TITLE
Change boundary condition generated by autodiff to a function update.

### PR DESCRIPTION
**Pull request by BachiLi**
The autodiff code currently generates two extra `Func`s for each adjoint function since it needs to take boundary condition into account when we transform scatter to gather. I propose to move the boundary condition to an update stage, so that 1) it makes the generated derivative code significantly more readable, and 2) it is easier to (auto)-schedule the gradient code since there are less `Func`s to consider.

### CODE REVIEW

**Comment by BachiLi** [comment]
Tests are probably failing for reasons outside of this PR. Seems like there are LLVM issues and some autoscheduler issues.

File: src/Derivative.cpp [hunk 1]
```diff
@@ -557,26 +557,39 @@ void ReverseAccumulationVisitor::propagate_adjoints(
...
+            }
             if (adjoint_func.values().size() == 1) {
-                Type type = adjoint_func.values()[0].type();
-                internal_assert(adjoint_func.function().output_types()[0] ==
-                                adjoint_func.values()[0].type());
-                Func f = BoundaryConditions::constant_exterior(
-                    adjoint_func, make_const(type, 0.0), box_to_vector(bounds));
-                adjoint_func = f;
-
+                Expr zero = make_const(adjoint_func.value().type(), 0.0);
```
Comment (abadams) [inline src/Derivative.cpp]: make_zero(adjoint_func.value().type())

**Comment by abadams** [comment]
The autoscheduler has an autoscheduled derivative pipeline with a checked-in machine-generated schedule, which is now invalid. That's why it's failing.

**Comment by BachiLi** [comment]
Cool. I'll try to fix it.

File: src/Derivative.cpp (hunk 1 above)
Comment (BachiLi) [reply src/Derivative.cpp]: There are too many make_const in Derivative.cpp. Fixed in https://github.com/halide/Halide/pull/4325

**Comment by BachiLi** [comment]
Fixed the autoscheduler cost model schedule by regenerating a schedule (needed to do some manual fixes due to some known bugs in the autoscheduler and image checks). As can be seen by the change the number of `Func` goes down from 96 to 56, which is a good sign.

**Comment by BachiLi** [comment]
Just pinging. Let me know what you think about this.

File: apps/autoscheduler/cost_model_schedule.h [hunk 2]
```diff
@@ -4,672 +4,922 @@
...
+inline void do_cost_model_schedule(Halide::Pipeline pipeline) {
+    // Manually removed all unrolls.
```
[the remark this answers is not in the document]
Comment (BachiLi) [reply apps/autoscheduler/cost_model_schedule.h]: Fixed.

File: src/Derivative.cpp [hunk 3]
```diff
@@ -557,26 +557,39 @@ void ReverseAccumulationVisitor::propagate_adjoints(
...
-                Type type = adjoint_func.values()[0].type();
-                internal_assert(adjoint_func.function().output_types()[0] ==
-                                adjoint_func.values()[0].type());
-                Func f = BoundaryConditions::constant_exterior(
-                    adjoint_func, make_zero(type), box_to_vector(bounds));
-                adjoint_func = f;
-
+                Expr zero = make_const(adjoint_func.value().type(), 0.0);
+                Expr value = adjoint_func(args);
+                Expr clamped_value = select(out_of_bounds, zero, value);
```
[the remark this answers is not in the document]
Comment (BachiLi) [reply src/Derivative.cpp]: Done. Why doesn't `constant_exterior` do it?

**Comment by steven-johnson** [comment]
arm32 and arm64 failures looks real

File: src/Derivative.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (steven-johnson) [reply src/Derivative.cpp]: Good question. @abadams @zvookin any idea why `constant_interior()` doesn't use `likely()`?

**Comment by BachiLi** [comment]
Hold on. I found an autodiff bug related to this and a simpler way to resolve that. Will open a new PR for that.

**Pull request closed without merging**